File: solutions/Leetcode_871/leetcode_871.cc
#include <queue>
#include <vector>

using namespace std;

class Solution1
{
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>> &stations)
    {
        priority_queue<int> q;
        q.push(startFuel);
        int idx = 0, distance = 0;
        int count = -1;
        while (distance < target && !q.empty())
        {
            distance += q.top();
            q.pop();
            count++;
            while (idx < stations.size() && distance >= stations[idx][0])
                q.push(stations[idx++][1]);
        }
        return distance >= target ? count : -1;
    }
};

class Solution2
{
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>> &stations)
    {
        vector<long> dp(stations.size() + 1, 0);
        dp[0] = startFuel;
        for (int i = 0; i < stations.size(); i++)
            for (int t = i; t > -1 && dp[t] >= stations[i][0]; t--)
                dp[t + 1] = max(dp[t + 1], dp[t] + stations[i][1]);
        for (int i = 0; i < dp.size(); i++)
            if (dp[i] >= target)
                return i;
        return -1;
    }
};

File: solutions/Leetcode_871/leetcode_871.py
import heapq
from typing import List


class Solution1:
    def min_refuel_stops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        distance = idx = 0
        count = -1
        fuels = [-startFuel]
        while distance < target and fuels:
            distance -= heapq.heappop(fuels)
            count += 1
            while idx < len(stations) and distance >= stations[idx][0]:
                heapq.heappush(fuels, -stations[idx][1])
                idx += 1
        return count if distance >= target else -1


class Solution2:
    def min_refuel_stops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        dp = [startFuel] + [0] * len(stations)
        for i in range(len(stations)):
            for t in range(i+1)[::-1]:
                if dp[t] >= stations[i][0]:
                    dp[t+1] = max(dp[t+1], dp[t] + stations[i][1])
        for i, x in enumerate(dp):
            if x >= target:
                return i
        return -1
